### PR TITLE
feat(auth): cloudchat-embedded JIT user provisioning

### PR DESCRIPTION
**Pull request by fjunqueira**
## Resumo

Provisiona o `User` row do typebot **just-in-time** quando um JWT Cognito válido chega pelo provider `cloudchat-embedded` do NextAuth. Resolve o erro **\"Failed to load flow builder. Please reload the page.\"** que aparece quando um user CloudChat acessa Eddie embedded pela primeira vez.

**Root cause** (pré-fix): o `authorize` do `cloudchat-embedded` fazia `prisma.user.findUnique({ where: { email } })` e retornava `null` se o user não existia. Como `CredentialsProvider` por design do NextAuth não invoca o adapter (`customAdapter.createUser`), users novos ficavam stuck.

**Workaround atual em prod** (Mat): pedir pro user autenticar manualmente via `custom-oauth` Cognito provider primeiro, que aciona `customAdapter.createUser` e popula o User row. Esse PR elimina a etapa manual.

## O que muda

**Cirúrgico no caminho embedded — `customAdapter` e `verifyCognitoToken` ficam intactos.**

| Arquivo | Mudança |
|---|---|
| `apps/builder/src/features/auth/helpers/isPrismaUniqueViolation.ts` (novo) | Type guard para `Prisma.PrismaClientKnownRequestError` com `code: 'P2002'` |
| `apps/builder/src/features/auth/helpers/createCloudChatEmbeddedUser.ts` (novo) | Helper minimal: cria User row + dispara telemetry `'User created'`. **Sem** workspace, **sem** apiToken, **sem** invitation processing |
| `apps/builder/src/features/auth/helpers/cloudchatEmbeddedAuthorize.ts` (novo) | Authorize callback extraído do inline em `[...nextauth].ts`. Lógica completa de JIT: validação de email, findUnique, create, P2002 race com refetch, logs estruturados |
| `apps/builder/src/pages/api/auth/[...nextauth].ts` | **+3/-47** linhas: provider passa a ser thin wrapper chamando o helper. Remove import de `verifyCognitoToken` (não usado mais nesse arquivo) |
| `*.test.ts` (3 novos) | 22 testes Vitest: 4 + 7 + 11. Cobre JIT happy/exists/race/refusal, email validation, field mapping, eddie_workspaces parsing |

**Pós-review (commit `578db80a`):** o `trackEvents` dentro do `createCloudChatEmbeddedUser` agora é wrapped em try/catch (telemetry é best-effort no caminho crítico de auth). Telemetry outage loga `warn 'cloudchat-embedded telemetry failed (user provisioned)'` em vez de surfacing como `JIT refused`. O teste original que validava propagação foi invertido pra validar o swallow. Spec doc + Datadog runbook atualizados.

## Decisões já bloqueadas no design

| | |
|---|---|
| Política de segurança | Qualquer JWT Cognito válido provisiona (paridade com OAuth flow). Barreira: `verifyCognitoToken` (issuer + audience + JWKS em prod). |
| `DISABLE_SIGNUP` | **NÃO** aplicado no JIT — flag é pra esconder botão de signup, não política de auth |
| Default workspace | **NÃO** criado — acesso vem do claim `custom:eddie_workspaces` (`listWorkspaces.getWorkspaceFilter` já tem `OR: [{ members: ... }, { id: { in: cognitoAccess.ids } }]`) |
| Default API token | **NÃO** criado — tokens são pra OAuth flow |
| Invitation processing | **NÃO** no JIT — convites typebot são conceito do produto standalone |
| Race condition | catch `P2002` + refetch (preserva atomicidade existente) |
| Observabilidade | Logs estruturados (5 níveis: error verify, warn missing-email, info JIT-provisioned, info race-resolved, warn JIT-refused) |

## Plan & spec docs

- [docs/superpowers/specs/2026-05-06-cloudchat-embedded-jit-provisioning-design.md](docs/superpowers/specs/2026-05-06-cloudchat-embedded-jit-provisioning-design.md) — design doc completo (decisões + rationale + dataflow + error matrix)
- [docs/superpowers/plans/2026-05-06-cloudchat-embedded-jit-provisioning.md](docs/superpowers/plans/2026-05-06-cloudchat-embedded-jit-provisioning.md) — plan de implementação task-a-task com TDD

## Como revisar

1. Diff é pequeno em código de produção (~150 LOC novas + 47 removidas em `[...nextauth].ts`)
2. Os 22 testes unit cobrem todos os branches da decisão (incluindo race + edge cases de claim parsing)
3. Para entender o porquê de cada decisão, ler o spec doc primeiro
4. Para entender o contrato exato de cada commit, ler o plan doc

## UAT local — antes de mergear

Como não temos staging, validação local + observabilidade pós-deploy. Roteiro completo no spec (\"Validation plan (local UAT)\") em 3 modos:

- **Modo 1**: CloudChat dev login + iframe Eddie real (testa caminho feliz E2E com `dev_cognito_token`)
- **Modo 2**: JWT craftado direto no callback (cobre `email_verified=true`, `name` setado, `hub_role: CLIENT`)
- **Modo 3**: 2 chamadas concorrentes (testa P2002 race com refetch resolved)

## Pós-deploy

Datadog query salva: `service:typebot-builder (\"JIT-provisioned\" OR \"JIT refused\" OR \"race resolved\" OR \"payload missing email\")`. Watch nos primeiros 30min do rollout pra confirmar que o JWKS real funciona com pessoas reais.

## Rollback

Branch é puramente aditivo no caminho embedded. Reverter restaura o `return null`. User rows já provisionados ficam no DB e funcionam no próximo deploy. Sem migração necessária.

## Follow-ups (intencionalmente fora de escopo)

- **`OAuthAccountNotLinked` no fluxo OAuth pós-JIT** (flagado pelo Cursor Bugbot, [#193 inline](https://github.com/cloudhumans/typebot.io/pull/193#discussion_r3196516818)) — user CloudChat-managed que JIT-provisionou via embed, se acessar Eddie standalone (`eddie.us-east-1.prd.cloudhumans.io` direto sem `?embedded=true`) e clicar "Cloud Hub Login", recebe erro porque NextAuth acha o User row mas sem `Account` linkado pro `custom-oauth` provider. Fix sugerido em PR separada: setar `allowDangerousEmailAccountLinking: true` no `custom-oauth` provider config — NextAuth passa a auto-linkar `Account` a `User` existentes quando email bate (seguro aqui porque o email é validado upstream pelo Cognito mesma User Pool). Pré-existe ao PR atual, mas o JIT torna o caso muito mais provável de acontecer.
- **`isNewUser=true` em todo login `cloudchat-embedded`** — o `authorize` retorna um `Pick<DatabaseUserWithCognito, ...>` sem `createdAt`, então o `signIn` callback (linha ~311 de `[...nextauth].ts`) avalia sempre como new user e dispara o disposable-email blocklist fetch pro GitHub. Pré-existe a esse PR (mesma shape no inline original), mas como o JIT path agora é exercitado de fato por mais users, fica mais visível. Opções: (a) incluir `createdAt` no return; (b) early-return no `signIn` callback pra `cloudchat-embedded`. Vou abrir issue separada.
- **Vitest como devDependency explícita** no workspace — atualmente os `*.test.ts` rodam via `npx vitest@4.1.5` (download on-demand) e não via `pnpm test`. Mesma situação dos testes pré-existentes (`accessControl.test.ts`, `cognitoUtils.test.ts`, etc.). Issue separada pra adicionar vitest config + script.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

Este PR extrai a lógica de autenticação `cloudchat-embedded` para um helper dedicado (`cloudchatEmbeddedAuthorize`) e adiciona provisionamento JIT (*just-in-time*) de usuários: quando um JWT Cognito válido chega pelo provider embedded e o e-mail não existe no banco, o helper cria o registro `User` automaticamente, lidando com race conditions via P2002 com refetch atômico.

- **`cloudchatEmbeddedAuthorize.ts`**: extrai o callback `authorize` do inline em `[...nextauth].ts`, adiciona JIT provisioning com tratamento de race P2002 e logs estruturados em 5 níveis; inclui `createdAt` no retorno, corrigindo o comportamento de `isNewUser` no `signIn` callback para usuários de retorno.
- **`createCloudChatEmbeddedUser.ts`**: helper mínimo que cria o registro `User` sem workspace/apiToken, com telemetria best-effort envolta em try/catch para não bloquear o caminho crítico de auth.
- **`isPrismaUniqueViolation.ts`** + 3 arquivos de teste: type guard reutilizável para `P2002` e 22 testes Vitest cobrindo happy path, race condition, edge cases de claim parsing e comportamento de telemetria.

<h3>Confidence Score: 5/5</h3>

Safe to merge. The change is surgical and additive: production auth behavior changes only for the `cloudchat-embedded` credentials provider, and the rollback path is a one-line revert with no migration needed.

The JIT provisioning logic is well-structured with correct P2002 race handling, telemetry isolated in best-effort try/catch, and `createdAt` forwarded so the `signIn` callback evaluates `isNewUser` correctly for returning users. The 22 unit tests cover all decision branches. Previously flagged issues (telemetry blocking auth, `isNewUser` always true) are both resolved in this commit. No blocking defects found in the changed code paths.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| apps/builder/src/features/auth/helpers/cloudchatEmbeddedAuthorize.ts | Core JIT authorize helper: token verification, findUnique, create, P2002 race handling, and structured logging. Correctly includes `createdAt` in the return type, fixing the `isNewUser` evaluation in the `signIn` callback. |
| apps/builder/src/features/auth/helpers/createCloudChatEmbeddedUser.ts | Minimal user creation helper with best-effort telemetry wrapped in try/catch. Correctly uses `?? undefined` for nullable fields to let Prisma use DB defaults. No side-effects (no workspace, no apiToken). |
| apps/builder/src/features/auth/helpers/isPrismaUniqueViolation.ts | Simple and correct type guard for Prisma P2002 unique constraint violations; well-tested. |
| apps/builder/src/pages/api/auth/[...nextauth].ts | Provider registration slimmed down significantly (-47 lines). The `cloudchat-embedded` provider is now a thin wrapper calling `cloudchatEmbeddedAuthorize`. `isNewUser` evaluation now works correctly for returning users since `createdAt` is included in the returned shape. |
| apps/builder/src/features/auth/helpers/cloudchatEmbeddedAuthorize.test.ts | 11 tests covering all decision branches: missing token, bad JWT, missing email, existing user, JIT happy path, P2002 race resolved, P2002+null refetch, non-P2002 error, full shape assertion, createdAt forwarding, email_verified mapping, and eddie_workspaces count parsing. |
| apps/builder/src/features/auth/helpers/createCloudChatEmbeddedUser.test.ts | 7 tests covering user creation with various field combinations, no-side-effects assertion, telemetry payload (first-name split), telemetry swallow on failure with warn log, and prisma error propagation. |
| apps/builder/src/features/auth/helpers/isPrismaUniqueViolation.test.ts | 4 tests covering true case (P2002), false case (non-P2002 code), plain Error, and non-Error values. Complete. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A([Embedded iframe login]) --> B[cloudchatEmbeddedAuthorize]
    B --> C{credentials.token present?}
    C -- No --> Z1([return null])
    C -- Yes --> D[verifyCognitoToken]
    D -- invalid --> Z2([return null, log error])
    D -- valid payload --> E{email in payload?}
    E -- No --> Z3([return null, log warn])
    E -- Yes --> F[DB findUnique by email]
    F -- found --> G([return AuthorizedUser])
    F -- not found --> H[createCloudChatEmbeddedUser]
    H --> I[p.user.create]
    I -- success --> J[trackEvents best-effort]
    J --> K([return AuthorizedUser, log JIT-provisioned])
    I -- P2002 race --> L[DB findUnique refetch]
    L -- found --> M([return AuthorizedUser, log race resolved])
    L -- null --> Z4([rethrow, null, log error])
    I -- other error --> Z5([return null, log warn JIT refused])
```

<sub>Reviews (8): Last reviewed commit: ["fix(auth): forward createdAt so signIn i..."](https://github.com/cloudhumans/typebot.io/commit/a1c9c99cdae2e713af344e03db4e327d2d5630d4) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=31029691)</sub>

### CODE REVIEW

**Comment by fjunqueira** [comment]
## ✅ Local UAT — passed (composezao stack)

Executei os 4 cenários do plano de validação local. **Todos passaram.**

### Modo 2 — JWT craftado com claims completos

JWT: `email=jit-uat-002@local.test`, `email_verified=true`, `name='Maria Cliente'`, `hub_role=CLIENT`, `eddie_workspaces='ws-real-1,ws-real-2'`.

Resultado:
- `POST /api/auth/callback/cloudchat-embedded` → HTTP 200 + session cookie
- DB: 1 User row, `name='Maria Cliente'`, `emailVerified=2026-05-06 17:59:45.384`
- 0 MemberInWorkspace, 0 ApiToken (✅ sem default workspace, sem default token)
- Log: `JIT-provisioned cloudchat-embedded user { userId, email, hubRole: 'CLIENT', eddieWorkspacesCount: 2 }`

### Modo 1-equivalente — dev_cognito_token shape

JWT igual ao que `dev_cognito_token()` emite no CloudChat dev (`apps/controllers/devise_overrides/sessions_controller.rb:442`): `email=jit-uat-001@local.test`, **sem** `email_verified`, **sem** `name`, `hub_role=ADMIN`, `eddie_workspaces='*'`.

Resultado:
- POST → HTTP 200
- DB: User row com `name=NULL`, `emailVerified=NULL` (✅ field mapping correto pra payload sem esses claims)
- 0 MemberInWorkspace, 0 ApiToken
- Log: `JIT-provisioned cloudchat-embedded user { userId, email, hubRole: 'ADMIN', eddieWorkspacesCount: 1 }` (✅ wildcard `'*'` count = 1, exatamente como o teste unit prediz)

### Modo 3 — race condition

6 callbacks concorrentes com mesmo JWT (`email=jit-uat-race@local.test`).

Resultado:
- Todos 6 → HTTP 200
- DB: **1 row** (✅ sem duplicação)
- Logs: `JIT-provisioned` (vencedor) + 5x `cloudchat-embedded JIT race resolved { email: 'jit-uat-race@local.test' }` (perdedores via P2002 → refetch)

### Edge case — payload sem email

JWT sem field `email` (só `sub` + `cognito:username`).

Resultado:
- POST → HTTP 401
- Log: `cloudchat-embedded payload missing email { sub: 'no-email', cognitoUsername: 'orphan' }` (✅ defesa em profundidade ativada)

### Modo 1 real — browser carregando o iframe Eddie

Naveguei via Playwright pra `http://localhost:3002/typebots?embedded=true&jwt=<dev_cognito_token shape>` com email novo (`jit-browser-uat@local.test`).

Resultado:
- Page title: **"claudia_project | Typebot"** (✅ NÃO o erro "Failed to load flow builder")
- 0 console errors
- UI carregou com workspace selector (`claudia_project Free`), 2 typebots seed (`Get Order Status`, `eddie-default`) visíveis
- DB pré: 0 rows; pós: 1 User row + 0 MemberInWorkspace + 0 ApiToken
- Log: `JIT-provisioned cloudchat-embedded user { ..., hubRole: 'ADMIN', eddieWorkspacesCount: 1 }`
- O user JIT-criado ganhou acesso visível ao workspace `claudia_project` via o claim `eddie_workspaces=*` → `getCognitoAccessibleWorkspaceIds` retorna `'admin'` → `listWorkspaces.getWorkspaceFilter` cai em `case 'admin'` → retorna todos workspaces não-pessoais. Funcionou exatamente como o design previa, **sem nenhuma row de MemberInWorkspace pro user JIT**.

### Logs gerais (todos os JIT-provisioned em ~3min de testes)

```
JIT-provisioned cloudchat-embedded user { userId, email, hubRole, eddieWorkspacesCount }  ✓ x 4 (todos os emails únicos)
cloudchat-embedded JIT race resolved { email }                                              ✓ x 5 (Modo 3)
cloudchat-embedded payload missing email { sub, cognitoUsername }                           ✓ x 1 (edge case)
cloudchat-embedded JIT refused                                                              ✓ x 0 (nenhum não-P2002 error nos testes)
Error in cloudchat-embedded authorize                                                       ✓ x 0
```

### Não-regressão do OAuth flow

`customAdapter.createUser` byte-for-byte intacto (verificado via `git diff origin/main..HEAD -- apps/builder/src/features/auth/api/customAdapter.ts` → empty diff).

---

Tudo verde, fora de draft.

File: docs/superpowers/specs/2026-05-06-cloudchat-embedded-jit-provisioning-design.md [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+| Payload missing/empty `email` | explicit check after verify | `return null` | `warn 'cloudchat-embedded payload missing email'` |
+| `findUnique` failure (DB issue) | catch externo (existing) | `return null` | `error` (same outer) |
+| User exists | `findUnique` returns user | `return user` | (silent; existing NextAuth jwt log fires) |
+| User does not exist, create succeeds | findUnique null + create OK | `return user` | `info 'JIT-provisioned cloudchat-embedded user' { userId, email, hubRole, eddieWorkspacesCount }` |
+| Race — P2002 caught, refetch returns user | inner catch | `return user` | `info 'cloudchat-embedded JIT race resolved'` |
+| Race — P2002 caught, refetch returns null | inner catch (rare/strange) | `throw` (caught by outer, returns null) | `error` outer |
+| Other Prisma error during create | inner catch | `return null` | `warn 'cloudchat-embedded JIT refused' { email, reason }` |
+| `trackEvents` throws (after user.create OK) | inner catch (non-P2002) | `return null`; user already in DB | `warn 'cloudchat-embedded JIT refused (telemetry)'` |
```
[the remark this answers is not in the document]
Comment (fjunqueira) [reply docs/superpowers/specs/2026-05-06-cloudchat-embedded-jit-provisioning-design.md]: Boa pegada. **Fixado em `578db80a`** — atualizei a tabela de error handling pra refletir o comportamento real do código (sem o `(telemetry)` suffix) e expandi o runbook do Datadog com a triage order: (1) DB issue, (2) claim shape change. Como bonus do fix do Cursor Bugbot ([#193 inline](https://github.com/cloudhumans/typebot.io/pull/193#discussion_r3196516818)), telemetry outage agora nem dispara `JIT refused` — vai como `cloudchat-embedded telemetry failed (user provisioned)` em log próprio porque `trackEvents` é swallowed dentro do `createCloudChatEmbeddedUser`.

File: docs/superpowers/specs/2026-05-06-cloudchat-embedded-jit-provisioning-design.md [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+export const createCloudChatEmbeddedUser = async ({
+  p,
+  email,
+  name,
+  emailVerified,
+  image,
+}: CreateCloudChatEmbeddedUserInput): Promise<User> => {
+  const user = await p.user.create({
+    data: { email, name, emailVerified, image, onboardingCategories: [] },
+  })
```
[the remark this answers is not in the document]
Comment (fjunqueira) [reply docs/superpowers/specs/2026-05-06-cloudchat-embedded-jit-provisioning-design.md:118]: Concern real e fora do escopo dessa PR — documentado como follow-up no [PR description](https://github.com/cloudhumans/typebot.io/pull/193) (seção "Follow-ups"). Cenário: user CloudChat-managed que JIT-provisionou via embed depois acessa Eddie standalone (`eddie.us-east-1.prd.cloudhumans.io` direto sem `?embedded=true`), clica "Cloud Hub Login" e dá erro `OAuthAccountNotLinked` porque NextAuth acha o User row mas sem Account linkado pro custom-oauth provider.

**Fix sugerido (PR separada):** setar `allowDangerousEmailAccountLinking: true` no `custom-oauth` provider config — NextAuth passa a auto-linkar Accounts a Users existentes quando o email bate. Faz sentido aqui porque o email é validado upstream pelo Cognito (mesma User Pool). Mas mexe num provider que essa PR prometeu não tocar, então fica pra próxima.

File: apps/builder/src/features/auth/helpers/createCloudChatEmbeddedUser.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  image?: string | null
+}
+
+export const createCloudChatEmbeddedUser = async ({
+  p,
+  email,
+  name,
+  emailVerified,
+  image,
+}: CreateCloudChatEmbeddedUserInput): Promise<User> => {
```
[the remark this answers is not in the document]
Comment (fjunqueira) [reply apps/builder/src/features/auth/helpers/createCloudChatEmbeddedUser.ts:19]: Esse review é do commit intermediário `33763483` (Task 2 — só os helpers). O wire-up aconteceu no commit seguinte `42a1a0cb` (Task 4) — `[...nextauth].ts:33` importa `cloudchatEmbeddedAuthorize` e `[...nextauth].ts:163` chama o helper como `authorize: (credentials) => cloudchatEmbeddedAuthorize(prisma, credentials ?? undefined)`. No HEAD da branch (`578db80a`) o caminho está completo. Fechar essa thread.

**Comment by fjunqueira** [comment]
## Endereçando o review summary do Greptile (Confidence 3/5)

@greptile-apps cobriu duas arestas no caminho crítico de auth. Endereçadas:

### Issue 1 — `trackEvents` propagation (FIXADO em `578db80a`)

Concern: se `trackEvents` falhasse depois do `prisma.user.create` ter sucesso, o erro propagava até `cloudchatEmbeddedAuthorize`, virava `JIT refused` e retornava `null` — mostrando "Failed to load flow builder" pro user mesmo com a User row já gravada. O teste `'propagates trackEvents errors (does not swallow)'` *validava* esse comportamento.

**Fix**: `createCloudChatEmbeddedUser` agora wrapping `trackEvents` em try/catch. Em falha de telemetria:
- User row continua persistida ✅
- Helper retorna o user normalmente ✅
- Loga `warn 'cloudchat-embedded telemetry failed (user provisioned)' { userId, email, error }`
- Auth path completa, user faz login ✅

Teste invertido: `'swallows trackEvents errors and returns the user (telemetry is best-effort)'`. Spec doc + Datadog runbook atualizados (a label `JIT refused` agora só dispara em falhas de DB/claim shape, não em telemetry hiccups).

### Issue 2 — `isNewUser=true` em todo login (DOCUMENTADO como follow-up)

Concern: o `AuthorizedUser` retornado pelo `authorize` é um `Pick<DatabaseUserWithCognito, ...>` sem `createdAt`. O callback `signIn` no NextAuth avalia `isNewUser = !('createdAt' in user && isDefined(user.createdAt))` — sempre `true` pra cloudchat-embedded → dispara fetch ao GitHub disposable-email blocklist em toda autenticação.

**Status**: pré-existe ao PR (mesma shape no `authorize` inline original). Documentado nos follow-ups do PR description com duas opções de fix: (a) incluir `createdAt` no return shape; (b) early-return no `signIn` callback pra esse provider. Não fixado aqui pra manter a PR cirúrgica.

---

Branch atualizada: `578db80a`. Tests 22/22 ✅. Lint/format clean. Pronto pra re-review.

**Review by giordanowt — CHANGES_REQUESTED** [review]
Revisei contra a análise que tínhamos feito do fluxo cloudchat-saas → claudia-app → typebot embedded. Resumo: **o fix bate exatamente na causa raiz** que diagnosticamos (`CredentialsProvider.authorize` retornando `null` por não invocar adapter ⇒ 401 ⇒ "Failed to load flow builder"), e está cirúrgico — `customAdapter.createUser` e `verifyCognitoToken` ficam intactos, cookies/CSP do iframe não mudam, race P2002 está bem tratada com refetch. Ótimo trabalho na separação dos helpers e na cobertura de testes.

Vou pedir changes por **um único motivo**, que listo como blocker abaixo. O resto são pontos de validação / follow-ups.

---

### 🔴 Blocker — `isNewUser=true` em todo login embedded vai bater no raw.githubusercontent.com no caminho crítico de auth

`apps/builder/src/pages/api/auth/[...nextauth].ts:354-363` calcula `isNewUser = !('createdAt' in user && isDefined(user.createdAt))` e, sendo `true`, faz `ky.get('https://raw.githubusercontent.com/.../disposable_email_blocklist.conf')` síncrono **antes** de aprovar o `signIn`.

O `authorize` em `cloudchatEmbeddedAuthorize.ts:67-78` devolve um `Pick<DatabaseUserWithCognito, ...>` sem `createdAt`. Resultado:

- A condição é `true` para **todo** login `cloudchat-embedded`, não só o JIT — inclusive users que já existiam há meses.
- Pré-PR isso afetava só os poucos clientes que tinham passado pelo workaround manual; pós-PR, **100% dos logins embedded** vão fazer um GET síncrono ao GitHub raw em produção.
- `raw.githubusercontent.com` não tem SLA, tem rate-limit anônimo (~60 req/h por IP) e já caiu mais de uma vez nos últimos 12 meses. Em incidente do GitHub ou sob carga, **auth do Eddie embedded inteiro quebra**.
- Bonus: se algum dia `DISABLE_SIGNUP=true` for usado como kill-switch, ele passa a bloquear logins embedded existentes (não só novos), porque `isNewUser` está sempre `true` — efeito não-intencional da flag.

Como mencionado no follow-up do próprio PR, fix é uma linha — incluir `createdAt` no retorno do `authorize` (e nos testes de mapping):

```diff
 type AuthorizedUser = Pick<
   DatabaseUserWithCognito,
   | 'id'
   | 'name'
   | 'email'
   | 'image'
   | 'emailVerified'
   | 'cognitoClaims'
   | 'cloudChatAuthorization'
   | 'cognitoTokenExp'
+  | 'createdAt'
 >
```

E no return final do `cloudchatEmbeddedAuthorize.ts`:
```diff
     return {
       id: user.id,
       email: user.email,
       name: user.name,
       image: user.image,
       emailVerified: user.emailVerified,
+      createdAt: user.createdAt,
       cognitoClaims: { ... },
       cloudChatAuthorization: true,
       cognitoTokenExp: payload.exp,
     }
```

Acho que vale tratar aqui mesmo (não como follow-up) porque é justamente o seu PR que torna esse caminho 100% exercitado em prod.

---

### 🟡 Pergunta — quem garante `custom:eddie_workspaces` ou `custom:hub_role` no Cognito para clientes novos?

`listWorkspaces.ts:34-35` lança `NOT_FOUND` se a query vier vazia. Pelo `getWorkspaceFilter`:

| `cognitoAccess.type` | Resultado para user JIT recém-provisionado |
|---|---|
| `admin` (`custom:hub_role=ADMIN`) | OK, vê todos workspaces non-personal |
| `restricted` (`custom:eddie_workspaces` setado) | OK, vê os workspaces listados |
| `none` (nenhum dos dois) | **`members: { some: { userId } }` → vazio → NOT_FOUND** |

O auth completa, JWT é emitido, mas a UX do iframe quebra com "No workspaces found". Vale confirmar: existe processo que setta esses claims no Cognito automaticamente quando um cliente novo entra no CloudChat? Se for setado pelo cloud-hub/backoffice no `PostConfirmation`, OK. Se depende de ação manual de admin, vamos ter um novo modo de falha pra clientes 100% novos. Não bloqueio por isso — só quero confirmar que está coberto.

---

### 🟡 Follow-up `OAuthAccountNotLinked` — concordo, mas merece ir junto

Você listou como follow-up separada e eu concordo com a análise. Só uma observação: o JIT torna esse caso **muito** mais provável (qualquer cliente novo do CloudChat agora pode tentar acessar Eddie standalone via "Cloud Hub Login" e tomar erro). `allowDangerousEmailAccountLinking: true` no `custom-oauth` provider é seguro porque a User Pool é a mesma e o email é validado upstream. Vale fazer no mesmo ciclo de deploy se possível, ou pelo menos abrir issue/PR já hoje.

---

### 🟢 Nit (cosmético, do Greptile, concordo)

`cloudchatEmbeddedAuthorize.test.ts:204` — depois do swallow do telemetry no `createCloudChatEmbeddedUser`, a mensagem mock `'telemetry endpoint down'` ficou enganosa. Trocar por `'db: connection lost'` deixa o teste autoexplicativo sobre qual categoria de erro chega ali agora.

---

### O que ficou bom

- Helpers separados e testáveis (👏 pelo `cloudchatEmbeddedAuthorize.ts` extraído em vez de ficar inline no `[...nextauth].ts`)
- Race P2002 + refetch
- Telemetry best-effort no commit `578db80a` (correto — auth não pode quebrar por falha de telemetria)
- Spec doc + plan doc explicando as decisões deliberadas (workspace default, DISABLE_SIGNUP, apiToken, invitations) — facilita muito a revisão
- Datadog query salva pra observar o rollout

Aprovando assim que o `createdAt` entrar.

File: apps/builder/src/features/auth/helpers/cloudchatEmbeddedAuthorize.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      name: user.name,
+      image: user.image,
+      emailVerified: user.emailVerified,
+      cognitoClaims: {
+        'custom:hub_role': payload['custom:hub_role'],
+        'custom:eddie_workspaces': payload['custom:eddie_workspaces'],
+      },
+      cloudChatAuthorization: true,
+      cognitoTokenExp: payload.exp,
+    }
```
[the remark this answers is not in the document]
Comment (fjunqueira) [reply apps/builder/src/features/auth/helpers/cloudchatEmbeddedAuthorize.ts:101]: Concern correto no código, mas **dormant em prod** — `DISABLE_SIGNUP` não está setada em nenhum dos configmaps em `typebot.io-manifests` (`grep -rIn DISABLE_SIGNUP /tmp/typebot.io-manifests/` retorna zero hits), então cai no default `false` do schema do `env.ts` e o gate nunca dispara. Mesma situação com `NEXT_PUBLIC_ONBOARDING_TYPEBOT_ID` (não setada → `newUser` redirect inativo).

Além disso, o shape `Pick<DatabaseUserWithCognito, ...>` sem `createdAt` é **pré-existente** ao PR — o `authorize` inline original tinha exatamente os mesmos campos. Esse PR não introduz nem piora a bomba latente.

Se um dia ligarem `DISABLE_SIGNUP=true`, vamos ter um problema **maior** do que a descrição (todos os logins cloudchat-embedded falhariam, não só os novos). Mas isso fica documentado como follow-up no [PR description](https://github.com/cloudhumans/typebot.io/pull/193) na mesma issue do `isNewUser=true` — o fix é uma única linha (incluir `createdAt` no Pick), endereça as duas concerns de uma vez.

Resolvendo a thread.

**Comment by fjunqueira** [comment]
@giordanowt valeu pelo review detalhado. Endereçando ponto a ponto:

### 🔴 Blocker — `createdAt` no Pick — **fixado em [`a1c9c99c`](https://github.com/cloudhumans/typebot.io/pull/193/commits/a1c9c99c)**

Você tá totalmente certo, e meu reasoning anterior (deixar como follow-up) tava errado. A regra "se n vai acontecer em prod n existe" que eu apliquei pro `DISABLE_SIGNUP` gate **não vale aqui** porque o GitHub fetch dispara em todo `isNewUser=true` sem env var como pré-requisito — ou seja, é ativo em prod hoje (volume baixo) e este PR explode pro 100% dos logins embedded.

Fix da one-liner:
- `cloudchatEmbeddedAuthorize.ts`: `'createdAt'` adicionado ao Pick + `createdAt: user.createdAt` no return (User row já tá carregada, sem custo extra)
- Comment inline explicando por que esse field tem que estar no return shape (futuro reviewer não cai na mesma armadilha)
- Bonus: o concern do Cursor sobre `DISABLE_SIGNUP` × `isNewUser=true` cai junto (mesmo root cause)
- Test: 22 → 23 (assertion explícita de que `createdAt` é forwardado pra returning users)

### 🟡 Pergunta — Cognito claims

Confirmação fora do meu alcance — verifiquei a árvore composezao com `grep -rIn 'custom:eddie_workspaces|custom:hub_role'` e só apareceram **leitores** (typebot, claudia-app, mcp-server, cloudchat-saas), nenhum writer. Os claims devem vir do Cognito User Pool via PostConfirmation Lambda ou do backoffice-api fora desse repo. Se você ou alguém do time souber o caminho exato, fico de boa em validar; caso contrário deixo a issue aberta como item de risco no PR description pra alguém com acesso ao backoffice cobrir.

### 🟡 OAuthAccountNotLinked — abrindo issue companion já

Concordo que `allowDangerousEmailAccountLinking: true` no `custom-oauth` provider é seguro (mesma User Pool, email validado upstream). Vou abrir issue separada agora pra rastrear, mas sem entrar nesse PR pra manter o foco do review (mexer no `custom-oauth` provider abre superfície a mais).

### 🟢 Nit — mock message — fixado em `a1c9c99c`

`'telemetry endpoint down'` virou `'db: connection lost'` no teste do non-P2002 path. Concordo que ficou enganoso depois do swallow do telemetry.

---

Branch: `a1c9c99c`. Tests 23/23 ✅. Lint/format clean. Pode dar nova passada quando puder.

**Review by giordanowt — APPROVED** [review]
LGTM 🚀

Re-revisei o commit `a1c9c99c` e os comentários inline.

**Blocker resolvido**:
- `createdAt` no `Pick<DatabaseUserWithCognito, ...>` e no return do `cloudchatEmbeddedAuthorize` (`apps/builder/src/features/auth/helpers/cloudchatEmbeddedAuthorize.ts`)
- Comentário inline no código explicando *por que* esse campo precisa ser forwardado (excelente — o próximo a ler isso vai entender em 5s)
- Teste explícito (`forwards user.createdAt so signIn callback's isNewUser check works for returning users`) — gostei de ter ficado como assertion separada em vez de só estar implícito no shape test
- Bonus: como você apontou, isso também mata o concern do `DISABLE_SIGNUP × isNewUser` do Cursor Bugbot (mesmo root cause). Eficiente.

**Nit do mock**: `'telemetry endpoint down'` → `'db: connection lost'` ✓

**Pontos amarelos respondidos**:
- *Cognito claims*: confirmei aqui também — `cloud-hub/src/utils/auth-helpers.ts:20` e `cloud-hub/src/types/user.ts:23` só leem o claim. A árvore que clonei (`cloud-hub`, `aws-access-control`, `cognito-user-list`, `backoffice-api`) não tem nenhum *writer*. Concordo com a hipótese PostConfirmation Lambda no User Pool ou backoffice-api fora dessa árvore. Vou tentar mapear o caminho exato e te confirmo no Slack — se for o caso, abrimos issue/checklist na issue companion.
- *`OAuthAccountNotLinked`*: concordo em manter escopo cirúrgico aqui. Como abriu issue companion, fica organizado.

**Antes do merge — só lembretes, não pedidos**:
- Rodar o UAT local nos 3 modos do spec ("Validation plan") com user real, não só dev token. O caminho do `email_verified` em prod (Cognito real) costuma ter pegadinhas.
- Confirmar que a Datadog query salva (`service:typebot-builder ("JIT-provisioned" OR "JIT refused" OR "race resolved" OR "payload missing email")`) está acessível pra quem for fazer o watch nos primeiros 30min do rollout.

Approve. Bom trabalho na separação dos helpers e na densidade dos testes — facilitou muito a revisão.

**Review by giordanowt — APPROVED** [review]
